File: backend/app/routers/asr.py
from fastapi import APIRouter, File, Form, HTTPException, UploadFile

from app.services.asr_service import ASRServiceError, transcribe_audio as transcribe_audio_file

router = APIRouter(prefix="/api/asr", tags=["asr"])


@router.post("/transcribe")
async def transcribe_audio(
    file: UploadFile = File(...),
    mode: str = Form("normal"),
) -> dict[str, object]:
    try:
        result = await transcribe_audio_file(file, mode)
    except ASRServiceError as error:
        raise HTTPException(status_code=error.status_code, detail=str(error)) from error

    return {
        "code": 200,
        "message": "success",
        "data": {
            "filename": file.filename,
            "mode": mode,
            "provider": result.provider,
            "raw_text": result.raw_text,
            "optimized_text": result.optimized_text,
        },
    }

File: frontend/src/api/asr.ts
export type TranscribeResponse = {
  code: number;
  message: string;
  data: {
    filename?: string;
    mode: string;
    raw_text: string;
    optimized_text: string;
  };
};

const API_BASE_URL = "http://127.0.0.1:8000";

type ApiErrorResponse = {
  detail?: string;
};

export async function transcribeAudio(audioBlob: Blob, mode = "normal") {
  const formData = new FormData();
  formData.append("file", audioBlob, "recording.webm");
  formData.append("mode", mode);

  const response = await fetch(`${API_BASE_URL}/api/asr/transcribe`, {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    let message = "上传音频失败，请确认后端服务已启动。";
    try {
      const errorData = (await response.json()) as ApiErrorResponse;
      if (errorData.detail) {
        message = errorData.detail;
      }
    } catch {
      // Keep the default message when the backend does not return JSON.
    }
    throw new Error(message);
  }

  return (await response.json()) as TranscribeResponse;
}
